fud2/fud-core/src/exec/driver.rs

```diff
@@ -132,40 +203,48 @@ impl Driver {
     pub fn plan(&self, req: Request) -> Option<Plan> {
         // Find a path through the states.
         let path =
-            self.find_path(req.start_state, req.end_state, &req.through)?;
-
-        let mut steps: Vec<(OpRef, Utf8PathBuf)> = vec![];
-
-        // Get the initial input filename and the stem to use to generate all intermediate filenames.
-        let (stdin, start_file) = match req.start_file {
-            Some(path) => (false, utils::relative_path(&path, &req.workdir)),
-            None => (true, "stdin".into()),
-        };
-        let stem = start_file.file_stem().unwrap();
+            self.find_path(&req.start_state, &req.end_state, &req.through)?;
 
         // Generate filenames for each step.
-        steps.extend(path.into_iter().map(|op| {
-            let filename = self.gen_name(stem, self.ops[op].output);
-            (op, filename)
-        }));
-
-        // If we have a specified output filename, use that instead of the generated one.
-        let stdout = if let Some(end_file) = req.end_file {
-            // TODO Can we just avoid generating the unused filename in the first place?
-            let last_step = steps.last_mut().expect("no steps");
-            last_step.1 = utils::relative_path(&end_file, &req.workdir);
-            false
-        } else {
-            // Print to stdout if the last state is a real (non-pseudo) state.
-            !self.states[req.end_state].is_pseudo()
-        };
+        let mut used_as_input = HashSet::new();
+        // collect filenames of outputs
+        let mut results = vec![];
```

Comment:
Even if it seems wasteful, this could perhaps be a lot easier to read if we got the result filenames in a separate pass, after generating all the filenames for all the ops.

Reply:
I agree!